Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,231 @@
+
+instance : Nonempty (@nullSubgroup M _) := ⟨0⟩
+
+variable (x : SeparationQuotient M)
+
+variable (z : M)
+
+/-- The norm of the image of `m : M` in the quotient by the null space is zero if and only if `m`
+belongs to the null space. -/
+theorem quotient_norm_eq_zero_iff (m : M) :
+    ‖mk m‖ = 0 ↔ m ∈ nullSubgroup := by
+  rw [norm_mk]
+  exact Eq.to_iff rfl
+
+/-- If for `(m : M)` it holds that `mk m = 0`, then `m  ∈ nullSubgroup`. -/
+theorem mk_eq_zero_iff (m : M) : mk m = 0 ↔ m ∈ nullSubgroup := by
+  rw [← quotient_norm_eq_zero_iff]
+  exact Iff.symm norm_eq_zero
+
+open NormedAddGroupHom
+
+/-- The morphism from a seminormed group to the quotient by the null space. -/
+noncomputable def normedMk : NormedAddGroupHom M (SeparationQuotient M) :=
+  { mkAddGroupHom with
+    bound' := ⟨1, fun m => by simp only [ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe,
+      mkAddGroupHom_apply, norm_mk, one_mul, le_refl]⟩}
+
+/-- `mkAddGroupHom` agrees with `QuotientAddGroup.mk`. -/
+@[simp]
+theorem normedMk.apply (m : M) : normedMk m = mk m := rfl
+
+/-- `normedMk` is surjective. -/
+theorem surjective_normedMk : Function.Surjective (@normedMk M _) :=
+  surjective_quot_mk _
+
+/-- The kernel of `normedMk` is `nullSubgroup`. -/
+theorem ker_normedMk : (@normedMk M _).ker = nullSubgroup := by
+  rw[ker, normedMk]
+  simp only [ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe, mk_toAddMonoidHom]
+  ext x
+  simp only [AddMonoidHom.mem_ker, AddMonoidHom.mk'_apply, mkAddGroupHom_apply]
+  exact mk_eq_zero_iff x
+
+/-- The operator norm of the projection is at most `1`. -/
+theorem norm_normedMk_le : ‖(@normedMk M _)‖ ≤ 1 :=
+  NormedAddGroupHom.opNorm_le_bound _ zero_le_one fun m => by simp only [normedMk.apply, norm_mk,
+    one_mul, le_refl]
+
+lemma eq_of_inseparable (f : NormedAddGroupHom M N) (hf : ∀ x ∈ nullSubgroup, f x = 0) :
+    ∀ x y, Inseparable x y → f x = f y := by
+  intro x y h
+  rw [inseparable_iff_norm_zero] at h
+  apply eq_of_sub_eq_zero
+  rw [← map_sub f x y]
+  exact hf (x - y) h
+
+/-- The lift of a group hom to the separation quotient as a group hom. -/
+noncomputable def liftNormedAddGroupHom (f : NormedAddGroupHom M N)
+    (hf : ∀ x ∈ nullSubgroup, f x = 0) : NormedAddGroupHom (SeparationQuotient M) N :=
+  { SeparationQuotient.liftContinuousAddMonoidHom ⟨f.toAddMonoidHom, f.continuous⟩
+      <| eq_of_inseparable f hf with
+    bound' := by
+      use ‖f‖
+      intro v
+      obtain ⟨v', hv'⟩ := surjective_mk v
+      rw [← hv']
+      simp only [ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe,
+        liftContinuousAddCommMonoidHom_apply, AddMonoidHom.coe_coe, norm_mk]
+      exact le_opNorm f v'}
+
+@[simp]
+theorem liftNormedAddGroupHom_apply (f : NormedAddGroupHom M N) (hf : ∀ x ∈ nullSubgroup, f x = 0)
+    (x : M) : liftNormedAddGroupHom f hf (mk x) = f x := rfl
+
+theorem norm_liftNormedAddGroupHom_apply_le (f : NormedAddGroupHom M N)
```

Comment:
Maybe docstrings on these, to help when people try to search for them via moogle, as moogle reads docstrings. (I might be wrong about this, though, so I defer to more experienced reviewers!)

Reply:
docstring added, but I'm not sure if the current `def`s are the right ones. that's why I added `RFC` tag